src/OrleansProviders/Streams/Memory/IMemoryStreamQueueGrain.cs

```diff
@@ -12,7 +12,7 @@ public interface IMemoryStreamQueueGrain : IGrainWithGuidKey
         /// <summary>
         /// Enqueue an event.
         /// </summary>
-        Task Enqueue(MemoryMessageData data);
+        Task Enqueue(MemoryMessageData data, int maxEventCount);
```

Comment:
This is a breaking-change. A new overload would be needed for this.

Reply:
Addressed.

Reply:
Unfortunately, it's still a breaking change: new silos will try to call this new method while it will not exist on older silos

Reply:
It's awful, but...:

the first 4 bytes of the ` Guid` key are a hash of the provider name. If you had a map `hash -> provider_name`, you could get the stream option that way...

Reply:
Or just have a global setting for all provider. I wonder if a lot of people are using different provider when using memory stream anyway?

Reply:
@benjaminpetit  True indeed. I can see a number of immediate options in `MemoryAdapterFactory.cs:182`.

1) Make an inline check on whether `MaxEventCount = DefaultMaxEventCount` and call the best method.
2) Also 1) but move the inline check to `MemoryStreamOptions.IsMaxEventCountDefault`, doing it once only. Can save a few nanos per call maybe.
3) Replace line 182 with a delegate call set upfront based on the check above. No branching or redundancy but probably worse perf than 2).

What do you suggest?

Reply:
If you really want, I think the best solution is to use a custom `MemoryAdapterFactory` implementation (this class isn't `sealed`) so you can implement your own implementation that will use some memory grain that are more friendly for your usage. You could use for example a compoundkey to inject the provider name in it.

Reply:
@benjaminpetit I took a stab at the hash map approach. To be fair, it's not any more clashy than what already exists. However you're correct, it's quite awful. Please tell me to revert it. Or at least to organise the key gen logic somewhere else. This looks very hacky.

I think having us write a new `MemoryAdapterFactory` on the app side would be overkill. It's not a trivial class and the provider works for us, we just need stuff not to be hardcoded.

Reply:
Sorry for the delay.

Yes, please revert it, my suggestion was awful. :')

I think the better way is to:
- extract methods from `IMemoryStreamQueueGrain` into a base interface
- make `MemoryAdapterFactory.GetQueueGrain` virtual and return the base interface introduced before
- optional: extract a base class for `MemoryStreamQueueGrain`

That way you could very easily write your own `MemoryAdapterFactory` implementation, that could call your own implementation of `IMemoryStreamQueueGrain` that could be configured.

Thought?